src/GraphQL.Tests/Types/CustomScalarTests.cs

```diff
@@ -0,0 +1,261 @@
+using System;
+using GraphQL.Language.AST;
+using GraphQL.SystemTextJson;
+using GraphQL.Types;
+using Shouldly;
+using Xunit;
+
+namespace GraphQL.Tests.Types
+{
+    public class CustomScalarTests : QueryTestBase<CustomScalarSchema>
+    {
+        [Theory]
+        [InlineData("externalNull", null)] // parsing "externalNull" should return null
+        [InlineData(null, "internalNull")] // parsing null should return "internalNull"
+        [InlineData("hello", "hello")]     // parsing "hello" should return "hello"
+        public void ParseLiteral(string externalValue, string internalValue)
+        {
+            var c = new CustomScalar();
+            // try parsing literal AST
+            c.ParseLiteral(externalValue == null ? (IValue)new NullValue() : new StringValue(externalValue)).ShouldBe(internalValue);
+        }
+
+        [Theory]
+        [InlineData("externalNull", null)] // parsing "externalNull" should return null
+        [InlineData(null, "internalNull")] // parsing null should return "internalNull"
+        [InlineData("hello", "hello")]     // parsing "hello" should return "hello"
+        public void ParseValue(string externalValue, string internalValue)
+        {
+            var c = new CustomScalar();
+            // try parsing value
+            c.ParseValue(externalValue).ShouldBe(internalValue);
+        }
+
+        [Theory]
+        [InlineData(null, "externalNull")] // serializing null should return "externalNull" (inverse of parsing)
+        [InlineData("internalNull", null)] // serializing "internalNull" should return null (inverse of parsing)
+        [InlineData("hello", "hello")]     // serializing "hello" should return "hello"
+        public void Serialize(string internalValue, string externalValue)
+        {
+            var c = new CustomScalar();
+            // try serializing
+            c.Serialize(internalValue).ShouldBe(externalValue);
+        }
+
+        [Theory]
+        [InlineData(null, "externalNull")] // serializing null should return "externalNull" (inverse of parsing)
```

Comment:
Are comments copy-pasted from Serialize?

Reply:
lol probably!

Reply:
let me look and see....

Reply:
ok, yes they are, but only because it's accurate, just as above where the parseliteral and parsevalue comments are the same

Reply:
I wrote a lot of comments in this file because I knew it would be hard to understand what was being tested, or why it worked the way it did.

Reply:
ok